File: easy/ruby/func_call.rb
require_relative 'common/vuln_child'
require_relative 'cmd_exec/interface_impl'

module FuncCall
  def self.run_by_child(cmd)
    child = VulnChild.new
    child.exec(cmd)
  end

  def self.run_by_interface(cmd)
    runner = InterfaceImpl.new
    runner.run(cmd)
  end
end

File: easy/c/func_call.c
#include "func_call.h"
#include "common/vuln_child.h"
#include "cmd_exec/interface_impl.h"

void run_by_child(const char* cmd) {
    VulnChild child = { {parent_exec} };
    child.parent.exec(cmd);
}

void run_by_interface(const char* cmd) {
    VulnChild child = { {parent_exec} };
    InterfaceImpl impl = { child, {impl_run} };
    impl.runner.run(&impl, cmd);
}
